### ISSUE TITLE
Build out well designed Disclaimer & Conditions of use screen

### ISSUE THREAD

**Issue by lukesleeman** [Western-Health-Covid19-Collaboration/wh_covid19_app#47]
The Disclaimer & Conditions of use screen needs to present a disclaimer.  It is accessible from two places:

1. When the user first starts the app it is the first thing they see
2. You can access it again from a button in the information screen

**We need to build build out a detailed version of the disclaimer & conditions of use screen**

@greggmiller Can you please provide some information in this issue on what should be in this section?  I think you had some copy?  Once that is done assign the issue to Marc Edwards.

@marcedwards Once greg has provided info, can you provide some design guidance in this issue in on how it should look.  Once this is done, can you label this issue as "Ready for Dev" so that one of the devs can pick it up.

**Comment by greggmiller** [comment]
This app is designed to be used by Anaesthetic and Critical Care doctors at Western Health who are caring for patients during the COVID-19 Pandemic. The guidelines, protocols and advice within this app are designed to educate staff and suggest recommended management of patients with COVID-19. The recommendations within this application do not indicate an exclusive course of action as the results must be contextualised to the patient at hand and variations may be appropriate. 
The authors of this app have made efforts to ensure the information is up to date, but encourage users to confirm the indicated therapy with existing guidelines. The responsibility is on you, the treating physician, to ensure the correct therapy is being used.

The authors accept no responsibility for any information perceived as misleading or the success of any treatment regimen detailed in the guidelines.
The app is made available on the understanding that Western Health and its employees shall have no liability (including liability by reason of negligence) to the users for any loss, damage, cost or expense incurred or arising by reason of any person using or relying on the information and whether caused by reason of any error, negligent act, omission or misinterpretation in the app or otherwise.

The Western Health trademark and app cannot be copied, modified, reproduced, published, uploaded, distributed or posted without the prior written consent of Western Health.

You must agree to accept these terms and conditions before proceeding to use the app.

**Comment by greggmiller** [comment]
^ Copy above and checkbox required at bottom.

**Comment by lukesleeman** [comment]
FYI just to unblock the developers I've opened up issue #73 to do a rough ugly version of the disclaimer screen.  This is just so we can get something into the app while the designers work on higher priority screens.

Continue to track the proper version of the screen though this issue 🙂

**Comment by lukesleeman** [comment]
Questions for @greggmiller :

1 - Could we have an "I accep" button instead of a checkbox?  An particular reason for a checkbox?

2 - Does this screen pop up _every_ time the user opens the app?  Or only the 1st time after they have installed it?

**Comment by greggmiller** [comment]
1. Yes! good idea
2. Just first time they use it. After that information is available in About section.

**Comment by marcedwards** [comment]
Here’s one approach we could take:

![Launch Disclaimer](https://user-images.githubusercontent.com/1121547/77501645-ba76ce80-6eac-11ea-8ac8-a6d360789ad0.png)

------

It’s very likely the text won’t all fit on one screen on most devices, and definitely won’t all fit on smaller devices. There’s a couple of solutions for that. If it’s okay for the person to accept the terms, we can just anchor the button to the bottom of the screen (also allowing for safe space for the home indicator etc).

![shorter](https://user-images.githubusercontent.com/1121547/77501740-04f84b00-6ead-11ea-8309-4fc6a6eead72.png)

------

If we need to be certain they’ve scrolled to read everything, we could just show “Please scroll down to agree” or similar, until they’re near the bottom of the text, then we could show the button.

![scroll-to-agree](https://user-images.githubusercontent.com/1121547/77501806-2fe29f00-6ead-11ea-9b46-b9fe23a2974b.png)

**Comment by benblackman** [comment]
Hi Ben here, another Anaesthetic doc at WH. 

I would have thought medicolegally it would make sense to ensure they've scrolled down as a proxy for ensuring they've read it. Doesn't really affect the usability of the app if we're only doing it on first use as per @greggmiller 's recommendation above. Thanks!

**Comment by marcedwards** [comment]
@benblackman Thanks! I agree that it’s not likely to be an issue, given it just happens once. We’ll take that path.

**Comment by marcedwards** [comment]
Just noting here that @Mattro and I discussed that “please scroll down to agree” is a bit ambiguous. Matt suggested “scroll to bottom to agree”. I think that’s better.

**Comment by lukesleeman** [comment]
@alessandrofavero  - I've assigned you to this issue!  Please update your current disclaimer work to match the designs.

I'm amazed how quickly this is coming together ⭐️

**Comment by marcedwards** [comment]
Here’s some sizes and specs that may be handy. Spacing and text is all very similar to the rest of the app. The spacing and margins is either 16pt or 32pt. I don’t know if we have a button component like this yet though?

The grey at the top just shows where the status bar is — we’d just have the normal status bar there, like the rest of the app.

![specs](https://user-images.githubusercontent.com/1121547/77519505-be6a1700-6ed3-11ea-8ee3-47cde30180b5.png)

------

The “Please scroll to bottom to agree” text should probably show until the _first time_ the user scrolls to the bottom, then it gets replaced with the “I Agree” button.

![scrolling](https://user-images.githubusercontent.com/1121547/77519621-e6f21100-6ed3-11ea-8faa-f9928dcb3ae3.gif)

**Comment by alessandrofavero** [comment]
@marcedwards What about something like this for the "scroll down" that disappear.
![Screenshot_1585127434](https://user-images.githubusercontent.com/43284065/77520148-da21ed00-6ed4-11ea-9cc7-441e0d3eed2f.png)

**Comment by marcedwards** [comment]
Sure! That looks great!

**Comment by greggmiller** [comment]
That looks great!

**Comment by alessandrofavero** [comment]
Feedback please @marcedwards 

First time disclaimer:
![intro_disclaimer](https://user-images.githubusercontent.com/43284065/77521842-69c89b00-6ed7-11ea-80a3-7d939ee7b11b.gif)

Disclaimer view from within the app:
![disclaimer_agreed](https://user-images.githubusercontent.com/43284065/77521884-777e2080-6ed7-11ea-83c1-770d966f54ee.gif)

**Comment by marcedwards** [comment]
Looking good! Some feedback:

- We probably don’t need to show the “Scroll down to agree” when they’ve agreed?
- There’s some new paragraphs in the text that don’t have a full paragraph space (maybe they need two return characters?).

**Comment by alessandrofavero** [comment]
> Looking good! Some feedback:
> 
> * We probably don’t need to show the “Scroll down to agree” when they’ve agreed?
> * There’s some new paragraphs in the text that don’t have a full paragraph space (maybe they need two return characters?).

No worries, I fixed them.

**Comment by benblackman** [comment]
Looking awesome, amazing work @alessandrofavero

**Comment by workerbee22** [comment]
Ok so we have a crash on iOS builds, but its a standard Flutter issue ie. Overflowing the bottom of the screen.

![image](https://user-images.githubusercontent.com/5154664/77707909-ae108400-701a-11ea-9505-8a79814701ac.png)

**Comment by workerbee22** [comment]
@alessandrofavero Also got this assertion failure on an Android Nexus 6P:

`
════════ Exception caught by widgets library ═══════════════════════════════════════════════════════
The following assertion was thrown building FutureBuilder<bool>(dirty, state: _FutureBuilderState<bool>#9dfcf):
'package:flutter/src/material/app_bar.dart': Failed assertion: line 199 pos 15: 'automaticallyImplyLeading != null': is not true.

The relevant error-causing widget was: 
  FutureBuilder<bool> file:///Users/gamma/AndroidStudioProjects/wh_covid19/wh_covid19_app/lib/view/disclaimer_view.dart:105:12
When the exception was thrown, this was the stack: 
#2      new AppBar (package:flutter/src/material/app_bar.dart:199:15)
#3      _DisclaimerViewState.build.<anonymous closure> (package:wh_covid19/view/disclaimer_view.dart:110:21)
#4      _FutureBuilderState.build (package:flutter/src/widgets/async.dart:732:48)
#5      StatefulElement.build (package:flutter/src/widgets/framework.dart:4619:28)
#6      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4502:15)
...
════════════════════════════════════════════════════════════════════════════════════════════════════
`

**Comment by alessandrofavero** [comment]
> @alessandrofavero Also got this assertion failure on an Android Nexus 6P:
> 
> `
> ════════ Exception caught by widgets library ═══════════════════════════════════════════════════════
> The following assertion was thrown building FutureBuilder(dirty, state: _FutureBuilderState#9dfcf):
> 'package:flutter/src/material/app_bar.dart': Failed assertion: line 199 pos 15: 'automaticallyImplyLeading != null': is not true.
> 
> The relevant error-causing widget was:
> FutureBuilder file:///Users/gamma/AndroidStudioProjects/wh_covid19/wh_covid19_app/lib/view/disclaimer_view.dart:105:12
> When the exception was thrown, this was the stack:
> #2 new AppBar (package:flutter/src/material/app_bar.dart:199:15)
> #3 _DisclaimerViewState.build. (package:wh_covid19/view/disclaimer_view.dart:110:21)
> #4 _FutureBuilderState.build (package:flutter/src/widgets/async.dart:732:48)
> #5 StatefulElement.build (package:flutter/src/widgets/framework.dart:4619:28)
> #6 ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4502:15)
> ...
> ════════════════════════════════════════════════════════════════════════════════════════════════════
> `

Thanks for the feedback @workerbee22 . I didn't encounter this error on testing (Pixel 4XL, Pixel 2, Pixel 3a, Galaxy Nexus). I think I know where the problem is, I need to handle the snapshot better. I'm at work now I will give it a look later today.

**Comment by alessandrofavero** [comment]
> Ok so we have a crash on iOS builds, but its a standard Flutter issue ie. Overflowing the bottom of the screen.
> 
> ![image](https://user-images.githubusercontent.com/5154664/77707909-ae108400-701a-11ea-9505-8a79814701ac.png)

@workerbee22 what is the resolution of the device that you used to test? Thanks

**Comment by workerbee22** [comment]
@alessandrofavero Overflow is on an iPhone 6 so not that high a resolution. Not sure the exact height, but we should always allow for almost any height?

**Comment by alessandrofavero** [comment]
> @alessandrofavero Overflow is on an iPhone 6 so not that high a resolution. Not sure the exact height, but we should always allow for almost any height?

@workerbee22  The thing is, based on the device height it either shows a column (no need to scroll, it fits in the view) or a ListView. I made that condition based on MediaQuery.of(context).size.height with a threshold at 600 logical pixel. I tested on Pixel 4XL, Pixel 3a, Pixel 2 and Galaxy Nexus and it worked just fine. By not having a Mac I'm not able to test on iOS unfortunately. We might need to adjust the threshold or find a different approach. Maybe
```
MediaQuery.of(context).size.height * MediaQuery.of(context).devicePixelRatio < 1080
```
Do you have a suggestion?

**Comment by workerbee22** [comment]
We should make it a scrollable widget regardless of height, but will have to take a closer look at the code tonight.

**Comment by domesticmouse** [comment]
Drop the media query and use [SingleChildScrollView](https://api.flutter.dev/flutter/widgets/SingleChildScrollView-class.html) instead. That way it'll do the right thing, no matter the context.

**Comment by alessandrofavero** [comment]
> Drop the media query and use [SingleChildScrollView](https://api.flutter.dev/flutter/widgets/SingleChildScrollView-class.html) instead. That way it'll do the right thing, no matter the context.

@domesticmouse true! But if I use a SingleChildScrollView how can I detect if the view fits or not the device display? If it doesn't fit (needs scrolling) I need to show the "Scroll down to agree".

**Comment by domesticmouse** [comment]
This sounds like a question for `StackOverflow`, as it is beyond my knowledge of the framework. There is a way to measure a widget, but you have to do it during the render phase if memory serves.

**Comment by alessandrofavero** [comment]
@domesticmouse  no worries! I'll see if any @Western-Health-Covid19-Collaboration/flutter-devs has a possible solution otherwise StackOverflow. Thanks!

**Comment by lukesleeman** [comment]
@alessandrofavero where did we get to with this issue?  I've seen the screen in the app now :-) Can we close this issue?  Or was there more work needed?

**Comment by alessandrofavero** [comment]
Hey @lukesleeman don't close it yet. I've created a new branch with some fixes and @workerbee22 is working on a different solution.

**Comment by workerbee22** [comment]
Still an issue ... should have a fix in today

**Comment by workerbee22** [comment]
@alessandrofavero which branch are you using for this? Or should I create a new one for a new PR? Was waiting for @joshcase to finish his fix first, which is [#106](https://github.com/Western-Health-Covid19-Collaboration/wh_covid19_app/pull/106) which is now just Merged by @lukesleeman

**Comment by workerbee22** [comment]
Note [issue #113](https://github.com/Western-Health-Covid19-Collaboration/wh_covid19_app/issues/113) is kind of related, so I'll be updating the structure of the app to support some Onboarding screens at the same time

**Comment by alessandrofavero** [comment]
> @alessandrofavero which branch are you using for this? Or should I create a new one for a new PR? Was waiting for @joshcase to finish his fix first, which is [#106](https://github.com/Western-Health-Covid19-Collaboration/wh_covid19_app/pull/106) which is now just Merged by @lukesleeman 

I was working on fix/disclaimer but feel free to open a new branch if you need

**Comment by workerbee22** [comment]
OK I'll open a new one and come back with you once I have the fix in.

**Comment by workerbee22** [comment]
@alessandrofavero , @maks and @joshcase just need to get some agreement on the fix for the Disclaimer screen as you have all been working on it or the `into_router`. Core requirements I has are:

- Disclaimer screen shown every time on startup until user presses I Agree button, after that it is never shown at app startup.
- Disclaimer must be able to be declined eg. by pressing Back in Android to close the app.
- Afer a user presses 'I Agree' button, app goes to Home screen.
- Once user has pressed 'I Agree' that first time, eg. pressing Back button on Android should not return to the Disclaimer screen for that session of the app or any other.

Any more?

**Comment by joshcase** [comment]
Yep that's my understanding of the goals, I am building dot point two as we speak

**Comment by maks** [comment]
Lets add a 'Decline' button to be explicit for the user that they don't accept disclaimer terms and so exist the app.

**Comment by workerbee22** [comment]
@joshcase I'm actually working on this already, nearly done, but is also a restructure to also allow future Onboarding screens.

**Comment by workerbee22** [comment]
Fix where disclaimer scrolls and button is fixed. Lots to cleanup here, text big to force being too long for screen etc. But you get the idea. And text will not go 'behind' the button etc. 

No MediaQuery and works on any density/height screen. Will cleanup tomorrow but wanted to post an update tonight to update you guys. @alessandrofavero , @maks and @joshcase 

Will also fix the other first run issues which I have a fix for too.

![disclaimer scrolling fixed button](https://user-images.githubusercontent.com/5154664/77846229-474cbf80-7200-11ea-89d6-c60191989886.gif)

**Comment by joshcase** [comment]
Brilliant, thanks @workerbee22. This was a real pain in the backside when running the project on my Android device.

**Comment by workerbee22** [comment]
Yeah ... its a nasty/tricky little issue and hard to figure out how to handle. But slightly better than being at the bottom and asking users to scroll @alessandrofavero because the button always on screen. Anyway ... will clean up in the morning and post a better GIF.

**Comment by workerbee22** [comment]
@maks will add Decline button too in the morning.

**Comment by alessandrofavero** [comment]
@workerbee22 great work looks good! Just one question, does the button become active when the user reaches the bottom of the screen or can be tapped even if the user didn't go fully trough the disclaimer?

**Comment by workerbee22** [comment]
@alessandrofavero Good question, at the moment it is always active, but can try to add that feature. I guess the trick is that we need to suggest to the user that the Disclaimer does go further down, like you had originally with screen shots above in this issue with the toast message indicating to scroll, if not all on screen. We can never really know the user actually read it, but your right we should at least indicate that 'it goes further down' in some way.

Also talking with some of the others added a requirement:
- When the Disclaimer text is changed AND the accompanying version flag, then the user is asked to re-accept the new Disclaimer ie. users must always have accepted the current Disclaimer text to use the app, as it may change.

This is easily tied into the on app start behaviour, so doing that this morning.

**Comment by workerbee22** [comment]
Nearly there team ... adding finishing touches.

**Comment by workerbee22** [comment]
@alessandrofavero @joshcase @maks and @sallychung  I've pushed to the `fix/disclaimer-update` branch, but won't create the PR until tested a little more. But everything is working on Android & iOS. 

Still need to pretty up the buttons back to what @alessandrofavero got them too. But wanted you guys to see the functional changes.

We can chat on Slack about how some of these things work, as there were some hard issues hidden in here with the disclaimer screen itself and the app startup/routes/navigator stuff.

**Comment by alessandrofavero** [comment]
@workerbee22 I'm checking the commits and looks very good, great job! BTW it missing when the user go to the disclaimer trough the info page (the disclaimer has already been agreed)

**Comment by workerbee22** [comment]
@alessandrofavero as per Slack, yeah I missed that will add back.

Also @Mattro discussions confirmed approach for iOS apps. Human Interface Guidelines from Apple prohibit an app closing itself, so some Flutter functions are disabled and won't close the app like they do with Android. 

So we've agreed to remove the new 'Decline' button @maks even on Android. Even though on Android you will still be able to hit Back button on the Disclaimer screen to exit the app.

**Comment by lukesleeman** [comment]
> So we've agreed to remove the new 'Decline' button @maks even on Android. Even though on Android you will still be able to hit Back button on the Disclaimer screen to exit the app.

Totally fine, I don't think I've ever seen a decline button in an app before.  Going back always does the job :-)

**Comment by maks** [comment]
@lukesleeman what's the plan for iOS though where no system back button? Just expect user to press home button?

**Comment by lukesleeman** [comment]
> @lukesleeman what's the plan for iOS though where no system back button? Just expect user to press home button?

🤷‍♂ Yep, I guess that works fine

**Comment by marcedwards** [comment]
Yep. If they don’t agree, then swiping home (or pressing home) and deleting the app is the action they’ll have to take. Totally fine, imo!

**Comment by sallychung** [comment]
Sorry if this has already been discussed but I didn’t see it in the comments here...

Are we keeping track of the user accepting the disclaimer and conditions of use based off a storing a version / date? It’d help ensure we cater for supporting changes to the T&Cs in the future if we need them to re-accept again.

I assume it’s a relatively small change to make before we release it out to beta users.

**Comment by marcedwards** [comment]
@sallychung That’s a great question and suggestion. Hopefully it’s easy to add. I think if we did need to make significant changes to the terms, we could reset it for everyone, but your suggestion would be a nicer way to go, if possible.

**Comment by workerbee22** [comment]
@alessandrofavero @maks @bramt I have pushed an update commit to `fix/disclaimer-update` branch. it has one bug, but everything else ... and there is heaps ... now fixed.

**Comment by workerbee22** [comment]
I'll raise a PR shortly after the Flutter event.

Note it also has the latest updates to the Disclaimer content.

**Comment by alessandrofavero** [comment]
I think we can close this one @workerbee22 ?

**Comment by workerbee22** [comment]
Yep I think so 👍Now that all functional and tests are in too.
